services/gitdiff/highlightdiff.go

```diff
@@ -0,0 +1,223 @@
+			tokenInMap = token + "<!-- " + tagStack[len(tagStack)-1] + "-->"
+			tagStack = tagStack[:len(tagStack)-1]
+		} else if token[0] == '<' { // for opening tag
+			tokenInMap = token
+			tagStack = append(tagStack, token)
+		} else if token[0] == '&' { // for html entity
+			tokenInMap = token
+		} // else: impossible
+
+		// remember the placeholder and token in the map
+		placeholder, ok := hcd.tokenPlaceholderMap[tokenInMap]
+		if !ok {
+			placeholder = hcd.nextPlaceholder()
+			if placeholder != 0 {
+				hcd.tokenPlaceholderMap[tokenInMap] = placeholder
+				hcd.placeholderTokenMap[placeholder] = tokenInMap
+			}
+		}
+
+		if placeholder != 0 {
+			res.WriteRune(placeholder) // use the placeholder to replace the token
+		} else {
+			// unfortunately, all private use runes has been exhausted, no more placeholder could be used, no more converting
+			// usually, the exhausting won't occur in real cases, the magnitude of used placeholders is not larger than that of the CSS classes outputted by chroma.
+			hcd.placeholderOverflowCount++
+			if strings.HasPrefix(token, "&") {
+				// when the token is a html entity, something must be outputted even if there is no placeholder.
+				res.WriteRune(0xFFFD)      // replacement character TODO: how to handle this case more gracefully?
+				res.WriteString(token[1:]) // still output the entity code part, otherwise there will be no diff result.
+			}
+		}
+	}
+
+	// write the remaining string
+	res.WriteString(htmlCode)
+	return res.String()
+}
+
+func (hcd *highlightCodeDiff) recoverOneDiff(diff *diffmatchpatch.Diff) {
+	sb := strings.Builder{}
+	var tagStack []string
+
+	for _, r := range diff.Text {
+		token, ok := hcd.placeholderTokenMap[r]
+		if !ok || token == "" {
+			sb.WriteRune(r) // if the rune is not a placeholder, write it as it is
+			continue
+		}
+		var tokenToRecover string
+		if strings.HasPrefix(token, "</") { // for closing tag
+			// only get the tag itself, ignore the trailing comment (for how the comment is generated, see the code in `convert` function)
+			tokenToRecover = token[:strings.IndexByte(token, '>')+1]
+			if len(tagStack) == 0 {
+				continue // if no opening tag in stack yet, skip the closing tag
+			}
+			tagStack = tagStack[:len(tagStack)-1]
+		} else if token[0] == '<' { // for opening tag
+			tokenToRecover = token
+			tagStack = append(tagStack, token)
+		} else if token[0] == '&' { // for html entity
+			tokenToRecover = token
+		} // else: impossible
+		sb.WriteString(tokenToRecover)
+	}
+
+	if len(tagStack) > 0 {
+		// close all opening tags
+		for i := len(tagStack) - 1; i >= 0; i-- {
+			tagToClose := tagStack[i]
+			// get the closing tag "</span>" from "<span class=...>" or "<span>"
+			pos := strings.IndexAny(tagToClose, " >")
+			if pos != -1 {
+				sb.WriteString("</" + tagToClose[1:pos] + ">")
+			} // else: impossible. every tag was pushed into the stack by the code above and is valid HTML opening tag
+		}
+	}
```

Comment:
This isn't right.

The return from `diffMatchPatch.DiffCleanupEfficiency` is `[]*Diff`.

Each `*Diff` represents an addition, unchanged or deletion section.

There is no guarantee that these sections have to be balanced.

The slice has to be balanced across the addition+unchanged or deletion+unchanged subsets all together.

Therefore you can have:

```
+ <span class="function">
- <span class="variable">
  amIaFunctionOrAVariable
+ ()
  </span>
```

Isn't this code going to result in:

```
+ <span class="function"></span>
- <span class="variable"></span>
  amIaFunctionOrAVariable
+ ()
  </span>
```

?

Reply:
In this case, the highlight is just lost, but the adding/removing tags are still correct. All tags are still balanced, and the last `</span>` won't be there, it should have been removed. https://github.com/go-gitea/gitea/pull/19958/files#diff-ee527c0c674cb10b21deadb5c08b523a8ff14727234ce3f752685c51d5e46b19R194-R200

I do not think there is a clear and easy way to make highlight still work perfectly in all cases, just better than before.

Reply:
See the demo in my last commit https://github.com/go-gitea/gitea/pull/19958/commits/9cf358b01a089fea78d476d1714265dda5f23204

It's guaranteed that tags in these sections are balanced individually.

Reply:
@zeripath ping. I think it is right. Do you have a real sample to break the code?